### ISSUE TITLE
Adjusting camera settings fails on macOS Monterey

### ISSUE THREAD

**Issue by naali** [openpnp/openpnp-capture#54]
Apparently macOS Monterey doesn't like opening & closing the USB control channel... 

I've noticed that similar issue was recently fixed for CameraController-project
[https://github.com/Itaybre/CameraController/commit/b8c7b47e64d79d55118db3e7461824ff4094836d](https://github.com/Itaybre/CameraController/commit/b8c7b47e64d79d55118db3e7461824ff4094836d)

```
svenn@MacBook-Pro tests % ./openpnp-capture-test 1
==============================
 OpenPNP Capture Test Program
 OSX 64 bit debug v0.0.22-6-gf9328d4 Mar 10 2022 
==============================
[INFO] Platform context created
2022-03-10 15:35:34.944 openpnp-capture-test[40580:20507881] Already have camera permission
[DBG ] enumerateDevices called
[DBG ] Name : FaceTime HD Camera (Apple Inc.)
[DBG ] Model: FaceTime HD Camera
[DBG ] U ID : FaceTime HD Camera (Apple Inc.) 47B4B64B70674B9CAD2BAE273A71F4B5
OSX Unable to extract vendor ID
OSX Unable to extract product ID
[DBG ] USB      : vid=0000  pid=0000
OSX Unique ID is not exactly 18 characters - wrong format to extract location.
We might have trouble identifying the UVC control interface.
[DBG ] Name : BRIO 4K Stream Edition (Unknown)
[DBG ] Model: UVC Camera VendorID_1133 ProductID_2155
[DBG ] U ID : BRIO 4K Stream Edition (Unknown) 0x2240000046d086b
[DBG ] USB      : vid=046D  pid=086B
OSX Unique ID is not exactly 18 characters - wrong format to extract location.
We might have trouble identifying the UVC control interface.
Number of devices: 2
ID 0 -> FaceTime HD Camera (Apple Inc.)
Unique:  FaceTime HD Camera (Apple Inc.) 47B4B64B70674B9CAD2BAE273A71F4B5
  Number of formats: 7
  Format ID 0: 1920 x 1080 pixels  30 FPS(max)  FOURCC=420v
  Format ID 1: 1280 x 720 pixels  30 FPS(max)  FOURCC=420v
  Format ID 2: 1080 x 1920 pixels  30 FPS(max)  FOURCC=420v
  Format ID 3: 1760 x 1328 pixels  30 FPS(max)  FOURCC=420v
  Format ID 4: 640 x 480 pixels  30 FPS(max)  FOURCC=420v
  Format ID 5: 1328 x 1760 pixels  30 FPS(max)  FOURCC=420v
  Format ID 6: 1552 x 1552 pixels  30 FPS(max)  FOURCC=420v
ID 1 -> BRIO 4K Stream Edition (Unknown)
Unique:  BRIO 4K Stream Edition (Unknown) 0x2240000046d086b
  Number of formats: 41
  Format ID 0: 160 x 120 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 1: 160 x 120 pixels  30 FPS(max)  FOURCC=420v
  Format ID 2: 176 x 144 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 3: 176 x 144 pixels  30 FPS(max)  FOURCC=420v
  Format ID 4: 320 x 180 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 5: 320 x 180 pixels  30 FPS(max)  FOURCC=420v
  Format ID 6: 320 x 240 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 7: 320 x 240 pixels  30 FPS(max)  FOURCC=420v
  Format ID 8: 340 x 340 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 9: 352 x 288 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 10: 352 x 288 pixels  30 FPS(max)  FOURCC=420v
  Format ID 11: 424 x 240 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 12: 424 x 240 pixels  30 FPS(max)  FOURCC=420v
  Format ID 13: 440 x 440 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 14: 480 x 270 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 15: 480 x 270 pixels  30 FPS(max)  FOURCC=420v
  Format ID 16: 640 x 360 pixels  30 FPS(max)  FOURCC=420v
  Format ID 17: 640 x 480 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 18: 640 x 480 pixels  30 FPS(max)  FOURCC=420v
  Format ID 19: 640 x 480 pixels  120 FPS(max)  FOURCC=420v
  Format ID 20: 800 x 448 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 21: 800 x 448 pixels  30 FPS(max)  FOURCC=420v
  Format ID 22: 800 x 600 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 23: 800 x 600 pixels  30 FPS(max)  FOURCC=420v
  Format ID 24: 848 x 480 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 25: 848 x 480 pixels  30 FPS(max)  FOURCC=420v
  Format ID 26: 960 x 540 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 27: 960 x 540 pixels  30 FPS(max)  FOURCC=420v
  Format ID 28: 1024 x 576 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 29: 1024 x 576 pixels  30 FPS(max)  FOURCC=420v
  Format ID 30: 1280 x 720 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 31: 1280 x 720 pixels  30 FPS(max)  FOURCC=420v
  Format ID 32: 1280 x 720 pixels  90 FPS(max)  FOURCC=420v
  Format ID 33: 1600 x 896 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 34: 1600 x 896 pixels  30 FPS(max)  FOURCC=420v
  Format ID 35: 1920 x 1080 pixels  30 FPS(max)  FOURCC=yuvs
  Format ID 36: 1920 x 1080 pixels  30 FPS(max)  FOURCC=420v
  Format ID 37: 1920 x 1080 pixels  60 FPS(max)  FOURCC=420v
  Format ID 38: 2560 x 1440 pixels  30 FPS(max)  FOURCC=420v
  Format ID 39: 3840 x 2160 pixels  30 FPS(max)  FOURCC=420v
  Format ID 40: 4096 x 2160 pixels  30 FPS(max)  FOURCC=420v
[DBG ] Setup for capture format (160 x 120)...
[DBG ] UVCCtrl::findDevice() called
[VERB] USB descriptor:
[VERB]   length    = 00000009
[VERB]   type      = 00000002
[VERB]   totalLen  = 00000CC0
[VERB]   interfaces = 00000005
[VERB] Configuration
[VERB] Interface[VERB]  VIDEO/CONTROL
[VERB] Processing Unit ID: 3
[DBG ] UVCCtrl::createControlInterface: created control interface
[DBG ] Created a UVC control object!
[DBG ] FOURCC = yuvs
Stream ID = 0
Stream is open
[INFO] PlatformStream::getPropertyLimits
[ERR ] sendControlRequest USBInterfaceOpen failed!
[VERB] getMinData failed
[ERR ] sendControlRequest USBInterfaceOpen failed!
[VERB] getMaxData failed
[ERR ] sendControlRequest USBInterfaceOpen failed!
[VERB] getDefault failed
Failed to get exposure limits!
[ERR ] sendControlRequest USBInterfaceOpen failed!
Failed to get exposure!
[VERB] UVCCtrl::getAutoProperty exposure
[ERR ] sendControlRequest USBInterfaceOpen failed!
Failed to get auto exposure!
Captured frames: 30

[INFO] closing stream
[DBG ] Stream::~Stream reports 30 frames captured.
[DBG ] Platform context destroyed
[DBG ] m_captureDevice released
[DBG ] m_captureDevice released
[DBG ] Context destroyed

```

**Comment by naali** [comment]
I created a pr which should fix this issue

https://github.com/openpnp/openpnp-capture/pull/55

**Comment by GatCode** [comment]
Thanks for creating a PR. Can you tell me how I can integrate your changes into openpnp? I need the adjustments but don’t want to downgrade in the meantime until your great improvements are integrated into the next openpnp release.

**Comment by naali** [comment]
Hi GatCode!

Just clone the repository as usual, ie (or just skip this step if you already have it cloned):
`git clone https://github.com/openpnp/openpnp-capture.git`

And then fetch the patch on top of it (run this under the directory where the project was cloned):
`git fetch origin pull/55/head:MASTER`

If you then check git log, you should see my changes on top of the list

```
svenn@MacBook-Pro openpnp-capture % git log    
commit dbaa7e75be6428f5f7fad47e785b65970fdf9adc (HEAD -> master)
Author: Antti Svenn <antti@pixieray.com>
Date:   Thu Mar 10 16:07:19 2022 +0200

    Don't open & close control channel on macOS Monterey

commit f9328d49c6af6fecfc329d3390b77efb9c79cb22 (origin/master, origin/HEAD)
Merge: 137f4e7 7553747
Author: Jason von Nieda <jason@vonnieda.org>
Date:   Mon May 17 14:36:50 2021 -0500

    Merge pull request #46 from openpnp/revert-45-arm64_support
    
    Revert "ARM64 support"

commit 755374706db64a475da46501d68d539d3a04798c (origin/revert-45-arm64_support)
Author: Jason von Nieda <jason@vonnieda.org>
Date:   Mon May 17 14:36:21 2021 -0500

    Revert "ARM64 support"
...
```

Then recompile and everything should be ok :)

**Comment by GatCode** [comment]
Hi @naali,
thanks for your reply and thank you very much for the explanation!
I was able to compile the lib fine but my question was a bit misleading - I want to know how I can recompile openpnp so that your changes are included in it. So that I can use openpnp on my pnp 😉
Sorry for the confusion!

**Comment by naali** [comment]
Ohhhh, I have no idea about the complete openpnp compilation. I've only been using this openpnp-capture library for totally different purposes.

**Comment by GatCode** [comment]
Oh ok 😉 no problem, thanks for your effort!

**Comment by vonnieda** [comment]
Fixed in #55, thanks @naali!
